File: p37_maze.cpp
#include <iostream>
#include <queue>

const int INF = 100000000;
const int N_MAX = 100, M_MAX = 100;

int N,M;
char field[M_MAX][N_MAX];

// to save shortest distances from start
int d[M_MAX][N_MAX];

// start point (x,y)
int sx, sy;
// goal point (x,y)
int gx, gy;

// vector for 4 direction
int dx[4] = {1, 0, -1, 0}, dy[4] = {0, 1, 0, -1};

typedef std::pair<int, int> P;

// breath-first search :　幅優先探索
int bfs(){

	std::queue<P> que;

	for( int y=0; y<N; y++)
		for( int x=0; x<M; x++)	d[x][y] = INF;
	
	// push start pos into que
	que.push(P(sx, sy));
	d[sx][sy] = 0;	// set start pos d = 0

	// if que is empty. it'll finish
	while(que.size()){

		P p = que.front(); que.pop();

		if (p.first == gx && p.second == gy) break;

		for (int i=0; i<4; i++){
			
			int nx = p.first + dx[i], ny = p.second + dy[i];

			if (nx>=0 && nx<M && ny>=0 && ny<N && field[nx][ny]!='#' && d[nx][ny]==INF){

				que.push(P(nx,ny));
				d[nx][ny] = d[p.first][p.second] + 1;
			}
		}

	}
	return d[gx][gy];

}

int main(){

	std::cin >> N >> M;

	for(int y=0; y<N; y++){
		for(int x=0; x<M; x++){
			std::cin >> field[x][y];
			if( field[x][y]== 'S'){
				sx = x;
				sy = y;	
			}else if(field[x][y] == 'G'){
				gx = x;
				gy = y;
			}
		}
	}
	std::cout << bfs() << "\n";
}

File: p37_maze.py
N = int( input() )
M = int( input() )
maze = [list( input() ) for i in range(N)]
INF = 10000000
pathcost = [[INF for i in range(M)] for j in range(N)]

sx = 0
sy = 0

for i in range(N):
    for j in range(M):
        if maze[i][j] == 'S':
            sx = j
            sy = i

dx = [1, 0, -1, 0]
dy = [0, 1, 0, -1]

def bfs():
    que = []
    que.append([sy,sx])
    pathcost[sy][sx] = 0
    while len(que) > 0:
        y, x = que.pop(0)

        for i, j in zip(dx, dy):
            ny = y + j
            nx = x + i
            if ny in range(N) and nx in range(M):
                if maze[ny][nx] == '.' and pathcost[ny][nx]==INF:
                    que.append([ny,nx])
                    pathcost[ny][nx] = pathcost[y][x]+1
                elif maze[ny][nx] == 'G':
                    pathcost[ny][nx] = pathcost[y][x]+1
                    return pathcost[ny][nx]

print(bfs())